web3/_utils/abi.py

```diff
@@ -302,6 +313,90 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
         return tuple()
 
 
+TUPLE_TYPE_STR_RE = re.compile(r'^(tuple)(\[([1-9][0-9]*)?\])?$')
+
+
+def get_tuple_type_str_parts(s: str) -> Optional[Tuple[str, Optional[str]]]:
+    """
+    Takes a JSON ABI type string.  For tuple type strings, returns the separated
+    prefix and array dimension parts.  For all other strings, returns ``None``.
+    """
+    match = TUPLE_TYPE_STR_RE.match(s)
+
+    if match is not None:
+        tuple_prefix = match.group(1)
+        tuple_dims = match.group(2)
+
+        return tuple_prefix, tuple_dims
+
+    return None
+
+
+def _align_abi_input(arg_abi, arg):
+    """
+    Aligns the values of any mapping at any level of nesting in ``arg``
+    according to the layout of the corresponding abi spec.
+    """
+    tuple_parts = get_tuple_type_str_parts(arg_abi['type'])
+
+    if tuple_parts is None:
+        # Arg is non-tuple.  Just return value.
+        return arg
+
+    tuple_prefix, tuple_dims = tuple_parts
+    if tuple_dims is None:
+        # Arg is non-list tuple.  Each sub arg in `arg` will be aligned
+        # according to its corresponding abi.
+        sub_abis = arg_abi['components']
+    else:
+        # Arg is list tuple.  A non-list version of its abi will be used to
+        # align each element in `arg`.
+        new_abi = copy.copy(arg_abi)
+        new_abi['type'] = tuple_prefix
+
+        sub_abis = itertools.repeat(new_abi)
+
+    if isinstance(arg, abc.Mapping):
+        # Arg is mapping.  Align values according to abi order.
+        arg = tuple(arg[abi['name']] for abi in sub_abis)
```

Comment:
We have a convention of avoiding overwriting local variables, preferring to instead use new variable names.  Maybe we can call this `aligned_arg`

```python
if is_mapping(arg):
    aligned_arg = tuple(...)
else:
    aligned_arg = arg
```

> NOTE: the above statement has an implicit `else` clause which is also somewhat against our conventions.  It might should look something like.

```python
if is_mapping(arg):
    ...
elif is_list_like(arg):
    ...
else:
    raise TypeError("Something something invalid")
```

This would simply be moving the `TypeError` from below up to be part of this clause.

Reply:
Also noting this relates to #1267 as it appears we already support dictionaries for the input arguments for tuple types.